### ISSUE TITLE
upload files and REST

### ISSUE THREAD

**Issue by intech** [moleculerjs/moleculer-web#54]
Multer to embed in the main code and display the parameters as done by bodyParser?
```js
...
routes: [
    {
        path: "/api",
            multer: {
               dest: "./uploads/",
               fields: [{ name: "logo", maxCount: 1}, { name: 'gallery', maxCount: 8 }],
            },
        aliases: {
            "REST sites": "sites",
        },
    },
...
]
```

In service `ctx.files.logo` and `ctx.files.gallery`

**Comment by icebob** [comment]
No, you have to do it [manually](https://github.com/moleculerjs/moleculer-web/blob/master/examples/full/index.js#L59)

**Comment by intech** [comment]
There is no way to access and manage files from the service. The only option where you can implement this in the `onBeforeCall`.

**Comment by tiaod** [comment]
Since 0.13 release, streaming response is possible, why not **streaming upload**?

**Comment by dothanhlam** [comment]
My working solution:

                path: "/api",
                aliases: {
              "POST /upload/file"(req, response) {
                        const form = new multiparty.Form();
                        const ref = this;
                        form.parse(req, function(err, fields, files) {
                            const file = files.file[0];
                            return ref.broker.call("amazon.uploadFile", {file})
                                .then(res => {
                                    ref.logger.info("File uploaded successfully!", res);
                                    response.end(res);
                                })
                                .catch(err => {
                                    ref.logger.error("File upload error!", err);
                                    ref.sendError(req, res, err);
                                });
                        });
                    },
                },

... and amazon service
actions: {
       uploadFile(ctx) {
            const { params } = ctx;
            const { file } = params;
            return this.uploadToS3({userId: 'knx33146', file});
        }
    },
}

 methods: {
        uploadToS3({userId, file}) {
            const buffer = fs.readFileSync(file.path);
            const key = `test-amazon-service/${file.originalFilename}`;

            const params = {
                Bucket: awsConfig.bucket,
                Key: key,
                Body: buffer,
            };

            return new Promise(function(resolve, reject) {
               s3.createBucket(function () {
                    s3.upload(params, function (err, data) {
                        if (err) {
                            console.log('error in callback');
                            console.log(err);
                        }
                        resolve(JSON.stringlify(data));
                    });
                })
            });
        },

**Comment by icebob** [comment]
Hi guys,
I've added a file upload implementation. Same as @dothanhlam's solution.

In route settings:
```js
{
    path: "/",

    bodyParsers: {
        json: false,
        urlencoded: false
    },

    aliases: {
        "FILE /upload": "file.save"
    },

    mappingPolicy: "restrict"
},
```

It calls the `file.save` action with the uploaded file stream. It uses the streaming function in Moleculer, so you can send the file to remote nodes, too.

I don't know this solution can cover all common use-cases. If you have free time, please test it and make suggestions if you have.

PR: #85

Install to try: `npm i moleculerjs/moleculer-web#file-upload`

**Comment by icebob** [comment]
I rewrote it, now it handles simple file streams, too. PR is refreshed: #85

**Comment by icebob** [comment]
Coming in 0.9.0

**Comment by Praful-Mali** [comment]
if  i want to upload file with input type than it will work ?

**Comment by icebob** [comment]
[Yes](https://github.com/moleculerjs/moleculer-web/blob/master/examples/full/assets/upload.html#L12)

**Comment by AhsanNissar** [comment]
I am trying to upload an excel file and then I need to write a function to perform custom validation on the file data. There can be 200-300 rows. I need to know how can i do this?

**Comment by DrJfrost** [comment]
the rpoblem with this type of approach is that it does not support array input files (many inputs with same input name), neither you can send multiple files with different names and an array of files (for example a `profile picture` with many image choices stored on a `choices` array)

**Comment by icebob** [comment]
Your request is not a common case and moleculer-web can't cover every situation. It just gives a common way to handle file uploading. If you have different and complex demand, you should write a custom alias and handle the request yourself.

**Comment by DrJfrost** [comment]
I trhink moleculer can improve if we have the hability to handle complex cases as well just like multer can handle all of these cases including single files via `single` method it can also handle complex functionality via `fields` method and even has a limiter for each of the fields defined. I would even consider migrating from busboy to multer which has more robust community and documentation :)

**Comment by DrJfrost** [comment]
## Proposal / Solution

I was able to implement file upload functionality for complex and normal cases with one of the cleanest solutions (if not the cleanest one), first it solved many issues:
1. Modification of body parsers was not used.
2. Configuration of route settings for busboy was not used.
3. Upload Functionality is implemented on Service side and not on APIGateway, following a nice microservice pattern.
4. `Req `and `Response` objects were sent to the service so the user could handle data easilly on the `req.body` and `req.params` options just like it does for express.
5. Multer was used instead of Busbuy, providing flexibility for file uploads.

## Metodology

First of all as @icebob said I implemented a custom alias to be able to handle the request by miself, however doing so opened the posibility to use Multer library.
The first challenge was to figure out how to send the `request` and `response` objects to the microservice. To do that I used a `$ctx` object that would allow me to access the context (AKA Broker) and also redirect `request` and `response` to the microservice like so:

```javascript
routes: [
	{
		path: "/api",

		whitelist: ["**"],

		//random settings here ...

		aliases: {
			async "POST v1/places/"(req, res){
				await req.$ctx.call(
					"v1.places.add",
					req, res
				).then((message)=>{
					if ("dataValues" in message){
						res.end(JSON.stringify(message));
					}else{
						res.end(message);
					}
				});
			},
		},

		// Calling options. More info: https://moleculer.services/docs/0.14/moleculer-web.html#Calling-options
		callingOptions: {
			retries: 3,
		},

		bodyParsers: {
			json: {
				strict: false,
				limit: "1MB",
			},
			urlencoded: {
				extended: true,
				limit: "1MB",
			},
		},

		// Mapping policy setting. More info: https://moleculer.services/docs/0.14/moleculer-web.html#Mapping-policy
		mappingPolicy: "all", // Available values: "all", "restrict"

		// Enable/disable logging
		logging: true,
	},
```

As you can see no necessary modification of `bodyparsers` was used. and the only thing I modified here was the custom Alias Behaviour. (I would like to note that it took me 8 hours to realize that req has a `$ctx` object becase there is literally no documentation about it, just a single line mentions it on the [Aliases Section](https://moleculer.services/docs/0.14/moleculer-web.html#Aliases). Once I realized I had a context it was a pain to realize how to programatically call actions from the gateway because once again there is no documentation about **how to programatically call services from the api gateway** however you guys can see I do that on the lines:

```javascript
await req.$ctx.call(
	"v1.places.add",
	req, res
)
```

Once I sent those things to the service I could handle that with Multer (including complex cases)

```javascript
const fsPromises = require("fs").promises;
const multer = require("multer");
const format = require("string-format");
const util = require("util");
const storage = require("../storage");
module.exports = {
	name: "places",
	version: 1,
	model: Place,
	actions: {
		add: {
			rest: {
				method: "POST",
				path: "/"
			},
			async handler({ params }) {
				return await this.add(params);
			},
		},
	},
	methods: {
		async add(req, res) {
			
			let message = null;
			try {
				const upload = util.promisify(storage.fields([
					{ name: "image", maxCount: 1 },
					{ name: "subplace_image", maxCount: 8 },
				]));
				await upload(req, res);


				
				// Insert your logic here after upoading files
				// you can also access req.body stuff and req.params as well



				return message //or Something_you_want_to_return;
			} catch (err) {
				if (err instanceof multer.MulterError) {
					// A Multer error occurred when uploading.
					if (err.code == "LIMIT_UNEXPECTED_FILE") {
						throw new Error(`Unexpected Field: ${err.field}`);
					} else {
						throw new Error(err.message);
					}
				} else if (err) {
					// An unknown error occurred when uploading.
					throw new Error(err.message);
				}
				message = err.message;
				return message;
			}
		},
	},
};


```

Finally I define a [disk storage](https://www.npmjs.com/package/multer#diskstorage) method to save files locally (this is the storage variable that I import in the previous file), you can change that for S3 file upload or any other service as well!


```javascript
const multer = require("multer");
const path = require("path");

var storage = multer.diskStorage({
  destination: function (req, file, cb) {
    cb(null, '/your/upload/path')
  },
  filename: function (req, file, cb) {
    cb(null, file.fieldname + '-' + Date.now())
  }
})

const upload = multer({ storage: storage);

module.exports = upload;
```
##Conclusion

With this approach you solve the problems described initially, also you have the ability to handle file upload and storage separately and also you have the ability to define service logic for it.  This approach can give you the hability to scale your project without affecting any body parsers or any other critical functionality of the entire project like the `apigateway` lastly it does not leave an open service for uploads that can be abused and exploited with different vulnerabilities like RFI (Remote File Inclusion) and/or (Local File Inclusion)

**Comment by icebob** [comment]
It can't work because you can't transfer the `req` and `res` object to a remote node. It works only if the API gateway and the `places` services are on the same node.

**Comment by DrJfrost** [comment]
oh I see what you mean. Is there a way to send `req` and `res` objects on remote nodes as well? for example saving those things on the req.$ctx.meta?, I really would love to have multer functionality on this since as you 
could see it greatly adds flexibility and scalability on the fileupload side and brings some serious use cases as well.

**Comment by icebob** [comment]
No, because they are not serializable. This is why we send the file content as a `Stream`. If you want to use multer, you should do it inside the API gateway.

**Comment by haris-reverie** [comment]
May be this comment would be helpful to those who are working on dockerized container.

I was stuck on something similar which needed transfering a file on multiple services, so what i did was in place of streaming or sending the file to different services i mounted all those services into a shared volume and stored that file into it.

Now I stored that file at the start of processing in the mounted volume directory, with which all the services out there could access or modify the stored file without need of passing the file to next service. In the end of processing task I just removed that file from volume to not keep on stacking junk files.
